### PR TITLE
feat: improve backups API

### PR DESCRIPTION
**Pull request by baptiste-grob**
Redo of https://github.com/standardnotes/snjs/pull/155

### CODE REVIEW

File: lib/application.ts [hunk 1]
```diff
@@ -930,19 +930,26 @@ export class SNApplication {
...
+    if (intent === EncryptionIntent.FileDecrypted) {
+      if (items.some(item => item.protected) && this.hasPasscode()) {
+        const passcode = this.challengeService.promptForPasscode(
+          ChallengeReason.CreateDecryptedBackupWithProtectedItems
+        );
+        if (!passcode) {
+          return;
+        }
+      }
+      items = items.filter(item => item.content_type !== ContentType.ItemsKey);
```
Comment (mobitar) [inline lib/application.ts:949]: Hmm, this is too smart for the application instance to know. Largely the application instance should just be a delegator. Why not have the protocol service do this logic?

Comment (baptiste-grob) [reply lib/application.ts:949]: I wanted the protocol service to not have to inspect its parameters and be more straightforward but upon reflection this isn't the right compromise either yes

File: lib/application.ts [hunk 2]
```diff
@@ -930,19 +930,25 @@ export class SNApplication {
...
-      intent,
-      returnIfEmpty
-    );
+    intent: EncryptionIntent
+  ): Promise<BackupFile | undefined> {
+    const items = this.itemManager.items;
+
+    if (intent === EncryptionIntent.FileDecrypted) {
+      if (items.some(item => item.protected) && this.hasPasscode()) {
+        const passcode = this.challengeService.promptForPasscode(
```
Comment (mobitar) [inline lib/application.ts:942]: Shouldn't this be awaited? And how is passcode being used? The return type for promptForPasscode isn't even a passcode, it's an object containing the passcode. Sounds like this functionality should be covered by a test.

File: lib/application.ts [hunk 3]
```diff
@@ -930,19 +930,25 @@ export class SNApplication {
...
+          ChallengeReason.CreateDecryptedBackupWithProtectedItems
+        );
+        if (!passcode) {
+          return;
+        }
+      }
+    }
+
+    return this.protocolService.createBackupFile(intent, items);
```
Comment (mobitar) [inline lib/application.ts:951]: Doesn't seem like we need to pass items here since they're unfiltered.

File: lib/application.ts (hunk 2 above)
Comment (mobitar) [reply lib/application.ts:942]: Oh I see, promptForPasscode is refactored in this PR. But still needs to be awaited.

**Comment by baptiste-grob** [comment]
Like I said in the previous PR:

> Heads up @radko93, this changes the method signature for createBackupFile(). It always returns an object now and you can decide how/if you want to JSON stringify it, add properties to it, etc.

File: lib/application.ts [hunk 4]
```diff
@@ -1193,7 +1203,7 @@ export class SNApplication {
         if (mergeLocal) {
           await this.syncService.markAllItemsAsNeedingSync(true);
         } else {
-          this.itemManager.removeAllItemsFromMemory();
+          void this.itemManager.removeAllItemsFromMemory();
```
Comment (mobitar) [inline lib/application.ts:1206]: Explain?

File: lib/services/protocol_service.ts [hunk 5]
```diff
@@ -736,41 +736,42 @@ export class SNProtocolService extends PureService implements EncryptionDelegate
...
+        }
+      })
+    );
+
+    const keyParams = await this.getRootKeyParams();
+
     const data: BackupFile = {
       version: this.getLatestVersion(),
-      items: encryptedPayloads.map((p) => p.ejected())
+      items: await ejectedPayloads,
```
Comment (mobitar) [inline lib/services/protocol_service.ts:769]: Why not `await` at the definition level instead? In this case the `ejectedPayloads` variable is misnamed, as it's a promise, not ejectedVariables. Seems like the await should be at line 747.

I don't think there's any performance optimization to be gained here either from postponing the await. JS is ultimately single-threaded.

File: test/backups.test.js [hunk 6]
```diff
@@ -25,29 +25,104 @@ describe('backups', () => {
     this.application = null;
   });
 
-  const BASE_ITEM_COUNT = 4;
+  const BASE_ITEM_COUNT_ENCRYPTED = [
+    'Note',
```
Comment (mobitar) [inline test/backups.test.js:29]: This is kind of weird. Here you're limiting all tests in this file to only deal with 2 notes. The idea with `BASE` is that these are the number of items that the system is creating that your test doesn't know about, so if you're counting items, make sure you account for these.

Comment (mobitar) [reply test/backups.test.js:29]: I don't think tests in this page should be limited to creating only 2 notes, and should be able to choose at their discretion how many notes to create. If they choose to create 10 notes, they can, and they should be able to ascertain that there are 10 + BASE items total.

File: lib/application.ts (hunk 4 above)
Comment (baptiste-grob) [reply lib/application.ts:1206]: `@typescript-eslint/no-floating-promises` will fire a warning when a promise is created but not awaited on or returned. This ensures that all async operations are accounted for.

In the case where we don't care to wait for the promise to return (fire-and-forget scenario), using the `void` operator (which does nothing) is used to silence the warning by explicitly communicating that we don't care about when the promise resolves or what its result is.

File: lib/services/protocol_service.ts (hunk 5 above)
Comment (baptiste-grob) [reply lib/services/protocol_service.ts:769]: @radko93 has reported notable performance improvements from "parallel" processing. 

While JS is single-threaded, when an async function is awaiting on a disk, network, or encryption result (like when you call into the RN bridge), it's put in a queue and other JS sync code can be run in the meantime. This is why using things like Promise.all, while not fully parallelizing the work, schedules it smarter, resulting in better performance. Does that make sense?

All that being said, in this instance `getRootKeyParams` doesn't seem to be a heavy operation, it just comes at the tail-end of a very heavy one (depending on how many notes/tags you have), so I went with as much smart scheduling as possible. If you think that makes it hard to read let me know and I'll `await` at the definition instead like you proposed.

File: test/backups.test.js (hunk 6 above)
Comment (baptiste-grob) [reply test/backups.test.js:29]: You're right, this is making too many assumptions about the tests in this file

Comment (baptiste-grob) [reply test/backups.test.js:29]: Are you saying the distinction between encrypted and decrypted base item count is weird?

Comment (mobitar) [reply test/backups.test.js]: > Are you also saying the distinction between encrypted and decrypted base item count is weird?

No, wasn't looking at that

**Review by mobitar — APPROVED** [review]

**Comment by baptiste-grob** [comment]
Watch out @radko93, `application.createBackupFile()` is slightly different as of 2.0.37 now :)

**Pull request merged**